### PR TITLE
Disable write/read Parquet when Parquet field IDs are used

### PR DESCRIPTION
**Pull request by res-life**
Temporarily disable write/read parquet when schema has a specified Parquet field ID
Solution is: Fallback to CPU when writing Parquet with field id column. Fallback to CPU if 
"spark.sql.parquet.fieldId.read.enabled" is true when reading Parquet.
Contributes #4846

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

File: sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/ParquetFieldIdShims.scala [hunk 1]
```diff
@@ -27,4 +31,18 @@ object ParquetFieldIdShims {
...
+    if (ParquetUtils.hasFieldIds(schema)) {
+      meta.willNotWorkOnGpu(
+        "Currently not support 'parquet.field.id' in parquet writer, schema is " + schema.json)
+    }
+  }
+
+  def tagGpuSupportReadForFieldId(meta: RapidsMeta[_, _, _], conf: RuntimeConfig): Unit = {
+    if(conf.get(SQLConf.PARQUET_FIELD_ID_READ_ENABLED.key, "false").toBoolean) {
+      meta.willNotWorkOnGpu("Currently not support reading field ids, " +
+          "please set spark.sql.parquet.fieldId.read.enabled as false")
```
Comment (jlowe) [inline sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/ParquetFieldIdShims.scala:45]:
```suggestion
      meta.willNotWorkOnGpu("reading by Parquet field ID is not supported, " +
          "${SQLConf.PARQUET_FIELD_ID_READ_ENABLED} is true")
```

File: sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/ParquetFieldIdShims.scala [hunk 2]
```diff
@@ -27,4 +31,18 @@ object ParquetFieldIdShims {
       SQLConf.PARQUET_FIELD_ID_WRITE_ENABLED.key,
       sqlConf.parquetFieldIdWriteEnabled.toString)
   }
+
+  def tagGpuSupportWriteForFieldId(meta: RapidsMeta[_, _, _], schema: StructType): Unit = {
+    if (ParquetUtils.hasFieldIds(schema)) {
```
Comment (jlowe) [inline sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/ParquetFieldIdShims.scala:36]: This check should not be performed unless `SQLConf.PARQUET_FIELD_ID_WRITE_ENABLED` is set to true.

File: tests/src/test/330+/scala/com/nvidia/spark/rapids/ParquetFieldIdSuite.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  test("try to write field id") {
+    val tmpFile = File.createTempFile("field-id", ".parquet")
+    try {
+      def withId(id: Int) =
+        new MetadataBuilder().putLong(ParquetUtils.FIELD_ID_METADATA_KEY, id).build()
+      // not support writing field id
+      val schema = new StructType().add("c1", IntegerType, nullable = true, withId(1))
+      val data = (1 to 4).map(i => Row(i))
+
+      assertThrows[IllegalArgumentException] {
```
Comment (jlowe) [inline tests/src/test/330+/scala/com/nvidia/spark/rapids/ParquetFieldIdSuite.scala:42]: This is not how we normally test for CPU fallbacks.  `IllegalArgumentException` could catch all sorts of exceptions that aren't actually the one we want/expect and then the test passes for the wrong reasons.  Typically we capture the plan and examine it to verify certain nodes are present and/or others are not present.

Also we prefer to put the fallback tests in the Python integration tests which have some nice framework to handle fallback testing.  See `assert_gpu_fallback_write` and the other fallback assertions in asserts.py.

As for the Parquet field ID metadata, you should be able to construct that in pyspark via something like this:
```python
from pyspark.sql.types import *

schema = StructType().add("c1", IntegerType(), nullable=True, metadata={ 'parquet.field.id' : 1 })
```

File: sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/ParquetFieldIdShims.scala (hunk 2 above)
Comment (res-life) [reply sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/ParquetFieldIdShims.scala]: done

File: tests/src/test/330+/scala/com/nvidia/spark/rapids/ParquetFieldIdSuite.scala (hunk 3 above)
Comment (res-life) [reply tests/src/test/330+/scala/com/nvidia/spark/rapids/ParquetFieldIdSuite.scala]: Done, moved to pyspark.

**Review by jlowe — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Comment by firestarman** [comment]
build

**Comment by sameerz** [comment]
build

**Pull request merged**